server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraCacheConfiguration.java

```diff
@@ -1,34 +1,78 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
 package org.apache.james.blob.cassandra.cache;
 
 import java.time.Duration;
-import java.time.temporal.ChronoUnit;
-
-import javax.inject.Inject;
-
-import org.apache.commons.configuration2.Configuration;
 
-import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
 
 public class CassandraCacheConfiguration {
 
-    private final String TIME_OUT_PROPERTY = "";
-    private final String MAX_SIZE_PROPERTY = "";
+    private final Duration timeOut;
+    private final int sizeThreshold;
+    private final int ttl;
 
-    private final long DEFAULT_TIME_OUT = 50;
-    private final int DEFAULT_MAX_SIZE = 8;
+    public CassandraCacheConfiguration(Duration timeout, int sizeThreshold, int ttl) {
+        this.timeOut = timeout;
+        this.sizeThreshold = sizeThreshold;
+        this.ttl = ttl;
+    }
 
-    final Duration TIME_OUT;
-    final int MAX_SIZE;
+    public Duration getTimeOut() {
+        return timeOut;
+    }
 
-    @Inject
-    public CassandraCacheConfiguration(Configuration propertiesConfiguration) {
-        this.TIME_OUT = Duration.of(propertiesConfiguration.getLong(TIME_OUT_PROPERTY, DEFAULT_TIME_OUT), ChronoUnit.MILLIS);
-        this.MAX_SIZE = propertiesConfiguration.getInt(MAX_SIZE_PROPERTY, DEFAULT_MAX_SIZE);
+    public int getTtl() {
+        return ttl;
     }
 
-    @VisibleForTesting
-    CassandraCacheConfiguration(long timeout, int maxSize) {
-        this.TIME_OUT = Duration.of(timeout, ChronoUnit.MILLIS);
-        this.MAX_SIZE = maxSize;
+    public int getSizeThreshold() {
+        return sizeThreshold;
+    }
+
+    public static class Builder {
```

Comment:
inner class go first, before fields

Reply:
This builder is subject to temporal coupling: if y forget to call some methods we are building an object with partially initialized values.

To avoid this we use `Optional<Type>` as fields of the builder, and either throw or supply default values in build method.

Reply:
IMO: we should move all validation in **CassandraCacheConfiguration()** then any property required missing should be throw. If we want to provide a default configuration, we should done it via injection properties. This way everything is simpler. right?

Reply:
In the past, the team already discussed the way to apply builder pattern in our code base.

The consensus is so far:

```
static class Builder {
   private A a; // a is compulsory, we could then check that it had been set in the build method. Optional is ok too.
   private Optional<B> b; // b is optional, thus having an optional allows setting the corresponding values to be set in the builder if needed
   private Boolean c; // c is compulsory, however as it is a raw type we need to use it's corresponding type to handle nullity. Optional is ok too.
   private Optional<Boolean> d; // D needs a default value hence we do like B
   // For collections use guava immutable builders

   private Builder() {
       // initialise optional / collection fields
       b = Optional.empty();
       d = Optional.empty();
   }

/*
    methods to mutate the state of the builder by realocating it's fields.
    We mutate the state of the collection builders without reaffecting them.
*/

   public MyObj build() {
       // Check that compulsory elements are here
       Precondition.checkState(a != null, "'a' is compulsory);
       Precondition.checkState(c != null, "'c' is compulsory);

       return new MyObj(
           a,
           b.orElse(B_DEFAULT_VALUE),
           c,
           d.orElse(D_DEFAULT_VALUE));
   }
}

// Private constructor to force using the builder
private MyObj(A a, B b, C c, D d) {
    // No validation needed as we know the builder was called to instanciate it
   this.a = a;
   this.b = b;
   this.c = c;
   this.d = d;
}
```

Would you agree with it?

Do you need help to apply it here?

Reply:
Well, i agree with it.